crates/nu-lsp/src/workspace.rs

```diff
@@ -0,0 +1,583 @@
+                if !content_string.contains(&text_to_search) {
+                    continue;
+                }
+                let temp_file = FullTextDocument::new("nu".to_string(), 0, content_string);
+                Self::find_reference_in_file(working_set, &temp_file, fp, &id)
+            }
+            .and_then(|range| self.occurrences.insert(uri, range));
+            self.send_progress_report(token.clone(), (i * 100 / len) as u32, None)?
+        }
+        self.send_progress_end(token.clone(), Some("Done".to_string()))
+    }
+
+    fn find_reference_in_file(
+        working_set: &mut StateWorkingSet,
+        file: &FullTextDocument,
+        fp: &Path,
+        id: &Id,
+    ) -> Option<Vec<Range>> {
+        let fp_str = fp.to_str()?;
+        let block = parse(
+            working_set,
+            Some(fp_str),
+            file.get_content(None).as_bytes(),
+            false,
+        );
+        let file_span = working_set.get_span_for_filename(fp_str)?;
+        let offset = file_span.start;
+        let mut references: Vec<Span> = find_reference_by_id(&block, working_set, id);
+
+        // NOTE: for arguments whose declaration is in a signature
+        // which is not covered in the AST
+        if let Id::Variable(vid) = id {
+            let decl_span = working_set.get_variable(*vid).declaration_span;
+            if file_span.contains_span(decl_span)
+                && decl_span.end > decl_span.start
+                && !references.contains(&decl_span)
+            {
+                references.push(decl_span);
+            }
+        }
+        let occurs: Vec<Range> = references
+            .iter()
+            .map(|span| span_to_range(span, file, offset))
+            .collect();
+
+        // add_block to avoid repeated parsing
+        working_set.add_block(block);
+        (!occurs.is_empty()).then_some(occurs)
+    }
+
+    fn get_workspace_folder_by_uri(&self, uri: &Uri) -> Option<WorkspaceFolder> {
+        let uri_string = uri.to_string();
+        self.workspace_folders
+            .iter()
+            .find_map(|(_, folder)| {
+                uri_string
+                    .starts_with(&folder.uri.to_string())
+                    .then_some(folder)
+            })
+            .cloned()
+    }
+
+    fn find_nu_scripts_in_folder(folder_uri: &Uri) -> Option<Vec<PathBuf>> {
+        let path = uri_to_path(folder_uri);
+        if !path.is_dir() {
+            return None;
+        }
+        let pattern = format!("{}/**/*.nu", path.to_string_lossy());
+        let (prefix, glob) = Glob::new(&pattern).ok()?.partition();
+        let path = canonicalize_with(&prefix, &path).ok()?;
+        Some(
+            glob.walk_with_behavior(
+                path,
+                WalkBehavior {
+                    // NOTE: as the progress is not cancellable ATM,
+                    // hard coded the max depth allowed in case it's triggered by accident
+                    depth: 5,
```

Comment:
I'm totally fine with a limitation like this initially but we need to remember to circle back because I could see how this could be a problem with just a few files but a lot of folders.